# Collatz-Vermutung

Daniel Naderer, 4CHIF, 22.09.2026

## Teil 1 – Die Folge erzeugen

Die Funktion erzeugt zu einer natürlichen Startzahl die vollständige Collatz-Folge. Die Startzahl und die abschließende 1 werden beide in die Liste aufgenommen. Die Rechenregel lautet:

$$
f(n) = 
\begin{cases}
\frac{n}{2} & \text{falls } n \text{ gerade ist} \\
3n + 1 & \text{falls } n \text{ ungerade ist}
\end{cases}
$$

`folge` speichert alle bisherigen Folgenglieder. `aktuelle_zahl` wird in jeder Schleifenrunde nach dieser Regel verändert, bis sie 1 erreicht.

In [ ]:
def collatz_folge(n):
    if not isinstance(n, int) or isinstance(n, bool) or n < 1:
        raise ValueError("Die Startzahl muss eine natürliche Zahl größer oder gleich 1 sein.")

    sequence = [n]
    curr = n

    while curr != 1:
        if curr % 2 == 0:
            curr //= 2
        else:
            curr = curr * 3 + 1
        sequence.append(curr)

    return sequence

Die Funktion wird nun mit drei unterschiedlichen Startzahlen getestet, darunter mit 27. Jede Ausgabe ist beschriftet, damit klar ist, zu welcher Startzahl die jeweilige Liste gehört.

In [ ]:
for i in (3, 27, 134):
    print(f"Collatz-Folge für {i}: {collatz_folge(i)}")

## Teil 2 – Rekordsuche

Nun werden alle Startzahlen von 1 bis 10 000 untersucht. Als **Schritte** zähle ich die Übergänge zwischen zwei Folgengliedern; deshalb hat eine Folge mit `schritte` Schritten genau `schritte + 1` Folgenglieder. Sobald eine Startzahl den bisherigen Rekord für die längste Folge überbietet, wird sie sofort ausgegeben.

In [15]:
long_start = None
long_step = -1
max_start = None
highest_max = -1
mehr_als_100_schritte = 0

for i in range(1, 10_001):
    seq = collatz_folge(i)
    steps = len(seq) - 1
    maximum = max(seq)

    if steps > long_step:
        long_start = i
        long_step = steps
        print(
            f"Neuer Rekord für die längste Folge: Startzahl {i}, "
            f"{steps} Schritte."
        )

    if maximum > highest_max:
        max_start = i
        highest_max = maximum

    if steps > 100:
        mehr_als_100_schritte += 1

print("\nErgebnisse der Rekordsuche:")
print(
    f"Die längste Folge beginnt bei {long_start} und braucht "
    f"{long_step} Schritte."
)
print(
    f"Das höchste Maximum erreicht die Startzahl {max_start}: "
    f"{highest_max:,}."
)
print(
    f"{mehr_als_100_schritte} von 10 000 Startzahlen brauchen mehr als 100 Schritte."
)

Neuer Rekord für die längste Folge: Startzahl 1, 0 Schritte.
Neuer Rekord für die längste Folge: Startzahl 2, 1 Schritte.
Neuer Rekord für die längste Folge: Startzahl 3, 7 Schritte.
Neuer Rekord für die längste Folge: Startzahl 6, 8 Schritte.
Neuer Rekord für die längste Folge: Startzahl 7, 16 Schritte.
Neuer Rekord für die längste Folge: Startzahl 9, 19 Schritte.
Neuer Rekord für die längste Folge: Startzahl 18, 20 Schritte.
Neuer Rekord für die längste Folge: Startzahl 25, 23 Schritte.
Neuer Rekord für die längste Folge: Startzahl 27, 111 Schritte.
Neuer Rekord für die längste Folge: Startzahl 54, 112 Schritte.
Neuer Rekord für die längste Folge: Startzahl 73, 115 Schritte.
Neuer Rekord für die längste Folge: Startzahl 97, 118 Schritte.
Neuer Rekord für die längste Folge: Startzahl 129, 121 Schritte.
Neuer Rekord für die längste Folge: Startzahl 171, 124 Schritte.
Neuer Rekord für die längste Folge: Startzahl 231, 127 Schritte.
Neuer Rekord für die längste Folge: Startzahl 313, 130 

## Teil 3 – Meine Beobachtungen

Für die Frage, ob größere Startzahlen auch längere Folgen haben, vergleiche ich einige unterschiedlich große Zahlen. Ausgegeben werden jeweils die benötigten Schritte und das höchste erreichte Folgenglied.

In [13]:
compare_num = (27, 100, 1_000, 5_000, 10_000)

for i in compare_num:
    seq = collatz_folge(i)
    print(
        f"Startzahl {i:>5}: {len(seq) - 1:>3} Schritte, "
        f"Maximum {max(seq):,}"
    )

Startzahl    27: 111 Schritte, Maximum 9,232
Startzahl   100:  25 Schritte, Maximum 100
Startzahl  1000: 111 Schritte, Maximum 9,232
Startzahl  5000:  28 Schritte, Maximum 5,000
Startzahl 10000:  29 Schritte, Maximum 10,000


### Fazit

**Die Folgen können sehr weit nach oben ausbrechen.** Die Startzahl 9 663 erreicht mit 27 114 424 das größte Maximum der Untersuchung. Das ist etwa 2 711-mal so groß wie die größte geprüfte Startzahl 10 000. Der Grund dafür ist die Regel für ungerade Zahlen: `3n + 1` vergrößert eine Zahl zunächst stark. Treffen danach nicht sofort genug gerade Zahlen aufeinander, können mehrere solche Vergrößerungen vorkommen, bevor die Halbierungen die Folge wieder senken. Für alle hier untersuchten Startzahlen endet die Folge zwar bei 1; das ist aber nur eine Beobachtung für diesen Bereich und kein Beweis der Vermutung.

**Verschiedene Startzahlen können denselben weiteren Weg haben.** Bei 6 ist der erste Schritt `6 → 3`. Ab der 3 läuft die Folge deshalb genau wie die Folge von 3 weiter: `3 → 10 → 5 → 16 → … → 1`. Das passiert, weil die nächste Zahl durch die Collatz-Regel eindeutig bestimmt ist. Sobald zwei Folgen dieselbe Zahl erreichen, sind alle folgenden Folgenglieder gleich.

**Eine größere Startzahl bedeutet nicht automatisch eine längere Folge.** Die Ausgabe zeigt zum Beispiel: 27 benötigt 111 Schritte, 10 000 aber nur 29 Schritte. Auch 5 000 kommt schon nach 28 Schritten bei 1 an. Die Startgröße allein sagt also wenig darüber aus, wie lang oder wie hoch eine Collatz-Folge wird.